### PR TITLE
adds debugger service to inspect properties of failed deployments

### PR DESCRIPTION
**Pull request by xh3b4sd**
Fixes https://github.com/giantswarm/giantswarm/issues/3290.

### CODE REVIEW

File: service/controller/v3/debugger/debugger.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/giantswarm/micrologger"
+
+	"github.com/giantswarm/azure-operator/service/controller/v3/key"
+)
+
+type Config struct {
+	Logger micrologger.Logger
+}
+
+type Debugger struct {
```
Comment (xh3b4sd) [inline service/controller/v3/debugger/debugger.go]: We introduce a new concept here so we can configure a debugger where it is needed so we can log more information we want to perceive in certain situations based on conditions that might be more complex.

File: service/controller/v3/debugger/debugger.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		return
+	}
+
+	d.logger.LogCtx(ctx,
+		"correlationID", *deployment.Properties.CorrelationID,
+		"id", *deployment.ID,
+		"level", "error",
+		"message", "deployment failed",
+		"name", *deployment.Name,
+	)
```
Comment (xh3b4sd) [inline service/controller/v3/debugger/debugger.go]: The fields here got suggested in the PR the original issue references.

File: service/controller/v3/resource/deployment/resource.go [hunk 3]
```diff
@@ -102,10 +109,10 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 	} else if err != nil {
 		return microerror.Mask(err)
 	} else {
-		s := *d.Properties.ProvisioningState
-		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment is in state '%s'", s))
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment is in state '%s'", *d.Properties.ProvisioningState))
 
-		if !key.IsFinalProvisioningState(s) {
+		if !key.IsFinalProvisioningState(*d.Properties.ProvisioningState) {
+			r.debugger.LogFailedDeployment(ctx, d)
```
Comment (xh3b4sd) [inline service/controller/v3/resource/deployment/resource.go]: Here we debug the failed deployment.

File: service/controller/v3/resource_set.go [hunk 4]
```diff
@@ -117,7 +130,8 @@ func NewResourceSet(config ResourceSetConfig) (*controller.ResourceSet, error) {
 	var deploymentResource controller.Resource
 	{
 		c := deployment.Config{
-			Logger: config.Logger,
+			Debugger: newDebugger,
```
Comment (xh3b4sd) [inline service/controller/v3/resource_set.go]: Here the deployment resource gets configured.

File: service/controller/v3/debugger/debugger.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (xh3b4sd) [reply service/controller/v3/debugger/debugger.go]: Can you elaborate? I am not sure what you mean or how this should be called instead.

[the remark this answers is not in the document]
Comment (kopiczko) [reply service/controller/v3/debugger/debugger.go]: I mean: `r.logger.LogCtx(ctx, r.debugger.LogFailedDeployment(d)...)`

[the remark this answers is not in the document]
Comment (xh3b4sd) [reply service/controller/v3/debugger/debugger.go]: Uncertain. Lets iterate.

**Pull request merged**